docker.nix

```diff
@@ -0,0 +1,264 @@
+{ pkgs ? import <nixpkgs> { }
+, lib ? pkgs.lib
+, name ? "nix"
+, tag ? "latest"
+, crossSystem ? null
+, channelName ? "nixpkgs"
+, channelURL ? "https://nixos.org/channels/nixpkgs-unstable"
+}:
+let
+  buildPkgs = pkgs;
+  targetPkgs =
+    if crossSystem != null && crossSystem != pkgs.system
+    then {
+      aarch64-linux = pkgs.pkgsCross.aarch64-multiplatform;
+      armv7l-linux = pkgs.pkgsCross.armv7l-hf-multiplatform.system;
+      x86_64-linux = pkgs.pkgsCross.gnu64;
+      powerpc64le-linux = pkgs.pkgsCross.musl-power;
+      i686-linux = pkgs.pkgsCross.gnu32;
+    }.${crossSystem}
+    else pkgs;
+
+  defaultPkgs = [
+    targetPkgs.nix
+    targetPkgs.bashInteractive
+    targetPkgs.coreutils-full
+    targetPkgs.gnutar
+    targetPkgs.gzip
+    targetPkgs.gnugrep
+    targetPkgs.which
+    targetPkgs.curl
+    targetPkgs.less
+    targetPkgs.wget
+    targetPkgs.man
+    targetPkgs.cacert.out
+    targetPkgs.findutils
+  ];
+
+  users = {
+
+    root = {
+      uid = 0;
+      shell = "/bin/bash";
+      home = "/root";
+      gid = 0;
+    };
+
+  } // lib.listToAttrs (
+    map
+      (
+        n: {
+          name = "nixbld${toString n}";
+          value = {
+            uid = 30000 + n;
+            gid = 30000;
+            groups = [ "nixbld" ];
+            description = "Nix build user ${toString n}";
+          };
+        }
+      )
+      (lib.lists.range 1 32)
```

Comment:
What is a reasonable number here? I have an AMD EPYC with 64 cores and 128 Hyperthreads.
Should this be configurable?

Reply:
This is the number used by NixOS so it seems reasonable here. Also, if you have lots of cores, you probably want parallelism *inside* a derivation rather than run a lot of derivations in parallel.

Hopefully this will become a non-issue with #3600.